### PR TITLE
#164798920 create search route

### PR DESCRIPTION
**Pull request by Chiazokam**
#### Description
This PR creates the search and filter functionality

#### Type of change
- [x] Task - non breaking change

#### How Has This Been Tested?
- [x] npm test
- [x] on postman

image

#### Checklist:
- [x] write tests for the endpoint
- [x] write validation for endpoint
- [x] create the endpoint

#### Story (PT)
[#164798920](https://www.pivotaltracker.com/story/show/164798920)

### CODE REVIEW

File: server/test/article-search.spec.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import chai, { expect } from 'chai';
+import chaiHttp from 'chai-http';
+import app from '../../index';
+
+
+chai.use(chaiHttp);
+describe('POST /api/v1/article/?query', () => {
```
Comment (davidokonji) [inline server/test/article-search.spec.js]: Please make corrections to the descriptions in the describe block, at `POST` should be `GET`

File: server/utils/articles/article-search.utils.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+class SearchValidators {
+/**
+ * @param {object} req
+ * @param {object} res
+ * @param {any} next
+ * @returns {any} response
+ */
+  static async checkQueryParams(req, res, next) {
+    const { author, title, tag } = req.query;
+    if (!author && !title && !tag) return Response(res, 422, 'No search parameters inputted');
```
Comment (davidokonji) [inline server/utils/articles/article-search.utils.js:12]: Please, can you explain what this validation does? I need clarification about this

File: server/utils/articles/article-search.utils.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    return next();
+  }
+
+  static async checkSpecialChars(req, res, next) {
+    const { author, title, tag } = req.query;
+    const checkRegex = /[`~!@#$%^&*()_|+\-=?;:'",.<>{}\\/]/;
```
Comment (davidokonji) [inline server/utils/articles/article-search.utils.js:18]: Please, what is the function of this regex expression?

File: server/test/article-search.spec.js (hunk 1 above)
Comment (Chiazokam) [reply server/test/article-search.spec.js]: Alright..thanks

File: server/utils/articles/article-search.utils.js (hunk 2 above)
Comment (Chiazokam) [reply server/utils/articles/article-search.utils.js:12]: At least one of the parameters should be present. If none of them is present, the user receives an error response

File: server/utils/articles/article-search.utils.js (hunk 3 above)
Comment (samorahno) [reply server/utils/articles/article-search.utils.js:18]: @davidokonji the regex expression is used to check for the presence of special characters, which in the long run prevents SQL injection.

File: server/test/article-search.spec.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import app from '../../index';
+
+
+chai.use(chaiHttp);
+describe('GET /api/v1/article/search', () => {
+  it('should return 200 for search query successfully run', async () => {
+    const res = await chai.request(app)
+      .get('/api/v1/article/search?author=Mofe&tag=&title=The%20boy%20drank%20palm&20wine');
+    expect(res).to.have.status(200);
+    expect(res.body.data.rows).to.be.an('array');
```
Comment (chukwudidu) [inline server/test/article-search.spec.js:12]: Good job testing for empty parameters and special characters. I think this test suite can be better improved by checking for the actual articles returned from the search. What do you think?

**Review by chukwudidu — CHANGES_REQUESTED** [review]

File: server/test/article-search.spec.js (hunk 4 above)
Comment (Chiazokam) [reply server/test/article-search.spec.js:12]: I think checking for the actual articles may not be the best idea. 
Tests may fail if there is a need later on to seed more articles.

**Review by geekdanitek — APPROVED** [review]

**Review by geekdanitek — COMMENTED** [review]
Kindly fix merge conflicts

File: server/test/article-search.spec.js (hunk 4 above)
Comment (Chiazokam) [reply server/test/article-search.spec.js:12]: I will implement this.

Comment (Chiazokam) [reply server/test/article-search.spec.js:12]: Done

**Review by chukwudidu — APPROVED** [review]

**Review by davidokonji — APPROVED** [review]

**Review by TheDrizzyWay — APPROVED** [review]
This looks good. Resolve your merge conflicts

**Review by BukkyOmo — APPROVED** [review]

**Review by samorahno — APPROVED** [review]
Looks good to me

**Review by paradisekelechi — APPROVED** [review]

**Pull request merged**